gdc_client/download/client.py

```diff
@@ -14,112 +15,87 @@
 
 class GDCDownloadMixin(object):
 
-    annotation_name = 'annotations.txt'
-
-    def download_related_files(self, index, file_id, directory):
-        """Finds and downloads files related to the primary entity.
-
-        :param str file_id: String containing the id of the primary entity
-        :param str directory: The primary entity's directory
+    def download_small_groups(self, smalls):
+        # List[List[str]] -> List[List[str]]
+        """Smalls are predetermined groupings of smaller filesize files.
+        They are grouped to reduce the number of open connections per download
 
         """
 
-        related_files = index.get_related_files(file_id)
-        if related_files:
-
-            log.info("Found {} related files for {}.".format(len(related_files), file_id))
-            for related_file in related_files:
-
-                log.debug("related file {}".format(related_file))
-                stream = DownloadStream(
-                    related_file, self.uri, directory, self.token)
-                self._download(self.n_procs, stream)
-
-                if os.path.isfile(stream.temp_path):
-                    utils.remove_partial_extension(stream.temp_path)
-        else:
-            log.debug("No related files")
-
-    def download_annotations(self, index, file_id, directory):
-        """Finds and downloads annotations related to the primary entity.
+        tarfile_url = self.data_uri + '?tarfile'
+        errors = []
+        groupings_len = len(smalls)
+        for i, s in enumerate(smalls):
+            try:
+                # post request
+                # {'ids': ['id1', 'id2'..., 'idn']}
+                ids = {"ids": s}
+
+                # using a POST request lets us avoid the MAX URL character length limit
+                r = requests.post(tarfile_url, stream=True, verify=self.verify, json=ids)
+                if r.status_code == requests.codes.ok:
+
+                    # {'content-disposition': 'filename=the_actual_filename.tar'}j
+                    filename = r.headers.get('content-disposition')
+                    filename = os.path.join(self.directory, filename.split('=')[1])
```

Comment:
If the response code is a 200, I don't think there is a possibility that there would be no filename. It's probably better to handle it. Too bad there's no :defense: emoji.